ciml/trainer.py

```diff
@@ -555,41 +596,49 @@ def local_trainer(dataset, experiment, gpu, debug):
     estimator = tf_trainer.get_estimator(
-        estimator, hyper_params, params, labels, model_dir)
+        estimator, hyper_params, params, labels, model_dir,
+        optimizer=_OPTIMIZER_CLS_NAMES[optimizer](learning_rate=learning_rate),
+        label_vocabulary=label_vocabulary)
 
-    # Now do the training and evalutation
-    if gpu:
-        with tf.device('/device:GPU:0'):
-            # Training
-            tf_trainer.get_training_method(estimator)(
-                input_fn=tf_trainer.get_input_fn(shuffle=True,
-                    batch_size=batch_size, num_epochs=num_epochs,
-                    labels=labels, **training_data), steps=steps)
-            # Eval
-            eval_loss = estimator.evaluate(
-                input_fn=lambda: tf_trainer.get_input_fn(
-                    batch_size=batch_size, num_epochs=1,
-                    labels=labels, **test_data))
-    else:
-        # Training
+    def train_and_eval():
+        # Train
         tf_trainer.get_training_method(estimator)(
             input_fn=tf_trainer.get_input_fn(shuffle=True,
                 batch_size=batch_size, num_epochs=num_epochs,
                 labels=labels, **training_data), steps=steps)
-        # Eval
-        eval_loss = estimator.evaluate(
-            input_fn=tf_trainer.get_input_fn(
-                batch_size=batch_size, num_epochs=1,
-                labels=labels, **test_data))
-    # Saving and Logging loss
-    print('Training loss after eval %r' % eval_loss)
-    eval_name = datetime.datetime.utcnow().strftime("%Y%m%d_%H%M") + "_eval"
-    gather_results.save_data_json(dataset, eval_loss, eval_name,
-                                  sub_folder=experiment)
-    gather_results.save_data_json(dataset, eval_loss, "eval",
-                                  sub_folder=experiment)
+        # Eval on the experiment dataset + any other requested
+        eval_sets = [dataset]
+        eval_sets.extend(eval_dataset)
+        for eval_dataset_name in eval_sets:
+            eval_data = gather_results.load_dataset(eval_dataset_name, 'test')
+            eval_size = len(eval_data['example_ids'])
+            print(
+                "Evaluation data shape: (%d, %d)" % eval_data['examples'].shape)
+            eval_loss = estimator.evaluate(
+                input_fn=tf_trainer.get_input_fn(
+                    batch_size=eval_size, num_epochs=1,
+                    labels=labels, **eval_data), name=eval_dataset_name)
+            # Saving and Logging loss
+            print('Training eval data for %s: %r' % (eval_dataset_name,
+                                                     eval_loss))
+            eval_name = "eval_" + eval_dataset_name
+            gather_results.save_data_json(dataset, eval_loss, eval_name,
+                                          sub_folder=experiment)
+
+    # Now do the training and evalutation
+    if gpu:
+        with tf.device('/device:GPU:0'):
```

Comment:
FWIW, when I was doing my local testing I don't think this really made a difference. The first gpu still wasn't being leveraged on my system. I realize this is just migrated from above, but just wanted to point that out as I went through everything. Although, maybe with the other changes tensorflow will decide that it can actually use the gpu now.